File: solutions/go/luhn/1/luhn.go
package luhn

import (
	"regexp"
)

var (
	reSpaces = regexp.MustCompile(`\s+`)
	reDigits = regexp.MustCompile(`\d+`)
)

// Valid function checks the given string for Luhn formula compliance.
func Valid(s string) bool {
	// remove spaces
	s = reSpaces.ReplaceAllLiteralString(s, "")
	if len(s) < 2 {
		return false
	}
	// remove all numerical digits
	everythingButTheDigits := reDigits.ReplaceAllLiteralString(s, "")
	if len(everythingButTheDigits) > 0 {
		return false
	}
	l := len(s) - 1
	parity := l % 2
	sum := int(s[l] - '0')
	for i := l - 1; i >= 0; i-- {
		d := int(s[i] - '0')
		if i%2 != parity {
			d *= 2
		}
		if d > 9 {
			d -= 9
		}
		sum += d
	}
	return sum%10 == 0
}

File: solutions/typescript/luhn/1/luhn.ts
const reSpaces = /\s+/g;
const reDigits = /\d+/g;

/**
 * Checks if the given string complies with the Luhn formula.
 * @param s The string to check.
 * @returns True if the string is valid according to the Luhn formula, false otherwise.
 */
export function valid(s: string): boolean {
    // Remove spaces
    s = s.replace(reSpaces, '');
    
    if (s.length < 2) {
        return false;
    }
    
    // Check if the string contains only digits
    if (s.replace(reDigits, '').length > 0) {
        return false;
    }
    
    const digits = s.split('').map(Number);
    const lastIndex = digits.length - 1;
    const parity = lastIndex % 2;
    
    const sum = digits.reduce((acc, digit, index) => {
        if (index === lastIndex) {
            return acc + digit;
        }
        
        if (index % 2 !== parity) {
            digit *= 2;
            if (digit > 9) {
                digit -= 9;
            }
        }
        
        return acc + digit;
    }, 0);
    
    return sum % 10 === 0;
}
